src/Interpreters/ProcessList.cpp

```diff
@@ -430,12 +437,37 @@ QueryStatus * ProcessList::tryGetProcessListElement(const String & current_query
 
 CancellationCode ProcessList::sendCancelToQuery(const String & current_query_id, const String & current_user, bool kill)
 {
-    auto lock = safeLock();
+    QueryStatus * elem;
+    CancelledProcesses::Iterator cancelled_it;
+
+    /// Cancelling the query should be done without the lock.
+    ///
+    /// Since it may be not that trivial, for example in case of distributed
+    /// queries it tries to cancel the query gracefully on shards and this can
+    /// take a while, so acquiring a lock during this time will lead to wait
+    /// all new queries for this cancellation.
+    ///
+    /// Another problem is that it can lead to a deadlock, because of
+    /// OvercommitTracker.
+    ///
+    /// So here the process is moved into cancelled.processes list, to wait
+    /// until it will be removed from it in ProcessListEntry dtor. And that way
+    /// we can call cancelQuery() without acquiring a lock.
+    {
+        auto lock = safeLock();
+        elem = tryGetProcessListElement(current_query_id, current_user);
+        if (!elem)
+            return CancellationCode::NotFound;
+        cancelled_it = cancelled.processes.insert(cancelled.processes.end(), elem);
+    }
 
-    QueryStatus * elem = tryGetProcessListElement(current_query_id, current_user);
+    SCOPE_EXIT({
+        DENY_ALLOCATIONS_IN_SCOPE;
 
-    if (!elem)
-        return CancellationCode::NotFound;
+        auto lock = unsafeLock();
```

Comment:
`unsafeLock` does not block overcommit tracker. On allocations are possible here, but there's a deallocation. Is it possible to enter overcommit tracker on deallocation?

Reply:
Actually yes, it is better to use `safeLock` here, indeed.